python/dgl/data/csv_dataset.py

```diff
@@ -0,0 +1,420 @@
+                dt = np.array([ast.literal_eval(row) for row in dt])
+            data[header] = dt
+        return data
+
+
+class DGLCSVDataset(DGLDataset):
+    """ This class aims to parse data from CSV files, construct DGLGraph
+        and behaves as a DGLDataset.
+
+    Parameters
+    ----------
+    data_path : str
+        Directory which contains 'meta.yaml' and CSV files
+    force_reload : bool
+        Whether to reload the dataset. Default: False
+    verbose: bool
+        Whether to print out progress information. Default: True.
+    data_parser : dict
+        A dictionary used for data parsing when loading data from CSV files.
+        The key is node/edge/graph type which specifies the header in CSV file
+        and the value is a callable object which is used to parse corresponding
+        column data. Especially, the graph level data parser is required
+        to be placed under key '_GRAPH_'. Default: None. If None, a default
+        data parser is applied which load data directly and tries to convert
+        list into array.
+
+    Attributes
+    ----------
+    graphs : :class:`dgl.DGLGraph`
+        Graphs of the dataset
+    data : dict
+        any available graph-level data such as graph-level feature, labels.
+
+    Examples
+    --------
+    Single heterograph dataset
+    >>> csv_dataset = data.DGLCSVDataset(test_dir)
+    >>> g = csv_dataset[0]
+
+    Multiple graphs dataset
+    >>> csv_dataset = data.DGLCSVDataset(test_dir)
+    >>> g, label = csv_dataset[0]
+    """
+    META_YAML_NAME = 'meta.yaml'
+
+    def __init__(self, data_path, force_reload=False, verbose=True, data_parser=None):
+        """
+        data_parser : dict
+            {'user' : NodeDataParser,
+             ('user', 'like', 'item') : EdgeDataParser,
+             '_GRAPH_' : GraphDataParser,
+            }
+        """
+        self.graphs = None
+        self.data = None
+        self.data_parser = {} if data_parser is None else data_parser
+        meta_yaml_path = os.path.join(data_path, DGLCSVDataset.META_YAML_NAME)
+        if not os.path.exists(meta_yaml_path):
+            raise DGLError(
+                "'{}' cannot be found under {}.".format(DGLCSVDataset.META_YAML_NAME, data_path))
+        self.meta_yaml = load_yaml_with_sanity_check(meta_yaml_path)
+        ds_name = self.meta_yaml.dataset_name
+        super().__init__(ds_name, raw_dir=os.path.dirname(
+            meta_yaml_path), force_reload=force_reload, verbose=verbose)
+
+    def process(self):
+        """Parse node/edge data from CSV files and construct DGL.Graphs
+        """
+        meta_yaml = self.meta_yaml
+        base_dir = self.raw_dir
+        node_data = []
+        for meta_node in meta_yaml.node_data:
+            if meta_node is None:
+                continue
+            ntype = meta_node.ntype
+            data_parser = DefaultDataParser() if ntype not in self.data_parser else self.data_parser[
+                ntype]
```

Comment:
Please initialize default parser in the `__init__` function.

Reply:
accepted.